File: Niveau 4/3 – Structures de données et Balayages/05 - Festival de musique.py
MAX_DAY = 100000

def main():
  nbJours, nGroupes = map(int, input().split())

  days = list(map(int , input().split()))

  dejaVu = [ 0 for _ in range(nGroupes + 1) ]
  groupeVue = 0

  m = MAX_DAY
  start = 0

  for end in range(nbJours):
    groupe = days[end]

    if dejaVu[groupe] == 0:
      groupeVue+=1
    dejaVu[groupe] += 1

    while groupeVue == nGroupes:
      m = min(m, end - start)
      dejaVu[ days[start] ] -=1
      if(dejaVu[days[start]] == 0):
        groupeVue -= 1
      
      start+=1
  
  print(m + 1)

main()

File: Niveau 4/3 – Structures de données et Balayages/05 - Festival de musique.cpp
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

const int MAX_DAY = 100000;

int main()
{
    ios::sync_with_stdio(false);

    int nbJours, nbGroupes;
    cin >> nbJours >> nbGroupes;

    vector<bool> groupesVue(nbGroupes, false);
    int nbGroupesVue = 0;

    vector<int> lastSeenGroup(nbGroupes, -1);

    int m = MAX_DAY;

    for (int i = 1; i <= nbJours; i++)
    {
        int groupe;
        cin >> groupe;
        groupe -= 1;

        lastSeenGroup[groupe] = i;

        if (nbGroupesVue != nbGroupes)
        {
            if (!groupesVue[groupe])
            {
                nbGroupesVue++;
                groupesVue[groupe] = true;
            }
        }
        if (nbGroupesVue == nbGroupes)
        {
            int minDay = MAX_DAY;
            int maxDay = 0;
            for (int element : lastSeenGroup)
            {
                minDay = min(minDay, element);
                maxDay = max(maxDay, element);
            }
            m = min(m, maxDay - minDay);
        }
    }

    cout << m + 1;
}